src/app/index.js

```diff
@@ -0,0 +1,234 @@
+import fsApi from 'fs';
+import path from 'path';
+import events from 'events';
+
+import config from 'config';
+
+import * as history from './history/index.js';
+import fetch from './fetcher/index.js';
+import filter from './filter/index.js';
+import loadServiceDeclarations from './loader/index.js';
+
+const fs = fsApi.promises;
+const __dirname = path.dirname(new URL(import.meta.url).pathname);
+const SERVICE_DECLARATIONS_PATH = path.resolve(__dirname, '../../', config.get('serviceDeclarationsPath'));
+
+export const AVAILABLE_EVENTS = [
+  'startTrackingChanges',
+  'endTrackingChanges',
+  'startRefiltering',
+  'endRefiltering',
+  'snapshotRecorded',
+  'firstSnapshotRecorded',
+  'noSnapshotChanges',
+  'versionRecorded',
+  'firstVersionRecorded',
+  'noVersionChanges',
+  'changesPublished',
+  'applicationError',
+  'documentUpdateError',
+  'documentFetchError',
+  'recordRefilterError',
+];
+
+export default class CGUs extends events.EventEmitter {
+  get serviceDeclarations() {
+    return this._serviceDeclarations;
+  }
+
+  async init() {
+    if (!this.initialized) {
+      this._serviceDeclarations = await loadServiceDeclarations(SERVICE_DECLARATIONS_PATH);
+      this.initialized = Promise.resolve();
+    }
+
+    return this.initialized;
+  }
+
+  attach(listener) {
+    AVAILABLE_EVENTS.forEach(event => {
+      const handlerName = `on${event[0].toUpperCase()}${event.substr(1)}`;
+
+      if (Object.prototype.hasOwnProperty.call(listener, handlerName)
+      || Object.prototype.hasOwnProperty.call(listener.constructor.prototype, handlerName)) {
+        this.on(event, listener[handlerName].bind(listener));
+      }
+    });
+  }
+
+  async trackChanges(serviceToTrack) {
+    try {
+      this.emit('startTrackingChanges');
+
+      const services = serviceToTrack ? { [serviceToTrack]: this._serviceDeclarations[serviceToTrack] } : this._serviceDeclarations;
+
+      const documentTrackingPromises = [];
+
+      Object.keys(services).forEach(serviceId => {
+        const { documents, name: serviceName } = this._serviceDeclarations[serviceId];
+        Object.keys(documents).forEach(type => {
+          documentTrackingPromises.push(this.trackDocumentChanges({
+            serviceId,
+            serviceName,
+            document: {
+              type,
+              ...documents[type]
+            }
+          }));
+        });
+      });
+
+      await Promise.all(documentTrackingPromises);
```

Comment:
We probably don't want to publish as soon as one document fails 🙊

> [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) rejects with the reason of the first promise that rejects.

```suggestion
      await Promise.allSettled(documentTrackingPromises);
```

Reply:
Yep, we have to do the same for refiltering